Zend/zend_compile.c

```diff
@@ -4712,6 +4712,169 @@ static void zend_compile_ns_call(znode *result, znode *name_node, zend_ast *args
 }
 /* }}} */
 
+static zend_op *zend_compile_rope_add(znode *result, uint32_t num, znode *elem_node);
+static zend_op *zend_compile_rope_add_ex(zend_op *opline, znode *result, uint32_t num, znode *elem_node);
+static void zend_compile_rope_finalize(znode *result, uint32_t j, zend_op *init_opline, zend_op *opline);
+
+static zend_result zend_compile_func_sprintf(znode *result, zend_ast_list *args) /* {{{ */
+{
+	if (args->children < 1) {
+		return FAILURE;
+	}
+
+	zend_eval_const_expr(&args->child[0]);
+	if (args->child[0]->kind != ZEND_AST_ZVAL) {
+		return FAILURE;
+	}
+
+	zval *format_string = zend_ast_get_zval(args->child[0]);
+	if (Z_TYPE_P(format_string) != IS_STRING) {
+		return FAILURE;
+	}
+	if (Z_STRLEN_P(format_string) >= 256) {
+		return FAILURE;
+	}
+
+	char *p;
+	char *end;
+	uint32_t string_placeholder_count;
+	
+	string_placeholder_count = 0;
+	p = Z_STRVAL_P(format_string);
+	end = p + Z_STRLEN_P(format_string);
+
+	for (;;) {
+		p = memchr(p, '%', end - p);
+		if (!p) {
+			break;
+		}
+
+		char *q = p + 1;
+		if (q == end) {
+			return FAILURE;
+		}
+
+		switch (*q) {
+		case 's':
+			string_placeholder_count++;
+			break;
+		case '%':
+			break;
+		default:
+			return FAILURE;
+		}
+
+		p = q;
+		p++;
+	}
+
+	/* Bail out if the number of placeholders does not match the number of values. */
+	if (string_placeholder_count != (args->children - 1)) {
+		return FAILURE;
+	}
+
+	/* Handle empty format strings. */
+	if (Z_STRLEN_P(format_string) == 0) {
+		result->op_type = IS_CONST;
+		ZVAL_EMPTY_STRING(&result->u.constant);
+
+		return SUCCESS;
+	}
+
+	znode *elements = NULL;
+
+	if (string_placeholder_count > 0) {
+		elements = safe_emalloc(sizeof(*elements), string_placeholder_count, 0);
+	}
+
+	/* Compile the value expressions first for error handling that is consistent
+	 * with a function call: Values that fail to convert to a string may emit errors.
+	 */
+	for (uint32_t i = 0; i < string_placeholder_count; i++) {
+		zend_compile_expr(elements + i, args->child[1 + i]);
+		if (elements[i].op_type == IS_CONST) {
+			if (Z_TYPE(elements[i].u.constant) != IS_ARRAY) {
+				convert_to_string(&elements[i].u.constant);
+			}
```

Comment:
You could potentially check if it is an object if it implements the `Stringable` interface? It would miss oddities like `GMP` not implementing the interface, although it can be cast to a string (which we really should fix...) but should work except if the `__toString()` method throws an error.

Reply:
These are compile time const expressions, they cannot currently contain objects (not even enums).